modules/geochemistry/include/utils/EquilibriumGeochemicalSystem.h

```diff
@@ -357,6 +363,218 @@ class EquilibriumGeochemicalSystem
+   *
+   * @param names the names of all the basis species, equilibrium species and surface-sorption
+   * minerals that are in the system.  Note, there must be exactly num_basis + num_eqm +
+   * num_surface_pot of these, ie, all species must be provided with a value
+   * @param values the values of the molalities (or solvent water mass for water, or free number of
+   * moles for minerals, or surface_potential_expr for surface-sorption sites).  There must be an
+   * equal number of these as "names".
+   * @param contraints_from_molalities whether the constraints initial provided to the
+   * EquilibriumGeochemicalSystem should be updated by the values provided.  This must have size
+   * num_basis.
+   *
+   * This method is reasonably nontrivial:
+   * - it assumes that temperature-dependent quantities (equilibrium constants, Debye-Huckel
+   * parameters, etc) have been set prior to calling this method, eg in the constructor or
+   * setTemperature()
+   * - it assumes the algebraic info has been built during instantiation, or during swaps, etc
+   * - it does not assume the value provided are "good", although since they most usually come from
+   * a previous solve, they are typically pretty "good".  It does check for non-negativity: molality
+   * for basis gases must be zero; molality for basis minerals must be non-negative; molality for
+   * every other basis species must be positive; molality for equilibrium gases and equilibrium
+   * minerals are set to zero irrespective of values provided; molality of other equilibrium species
+   * must be non-negative; surface_potential_expressions must be positive.  (In the previous
+   * sentence, "molality" is molality, kg solvent water, or free mineral moles, whichever is
+   * appropriate.)
+   * - if constraints_from_molalities is false then: if the original constraint was
+   * KG_SOLVENT_WATER, FREE_MOLALITY or FREE_MOLES_MINERAL_SPECIES then the constraint take
+   * precedence over the molality provided to this method, so the molality is ignored and the
+   * constraint value used instead.
+   * - if constraints_from_molalities is true then: if the original constraint was KG_SOLVENT_WATER,
+   * FREE_MOLALITY or FREE_MOLES_MINERAL_SPECIES then the constraint value is set to the value
+   * provided by to this method; if the original constraint was MOLES_BULK_WATER or
+   * MOLES_BULK_SPECIES then the constraint value is set to the value computed from the molalities
+   * provided to this method; if the original constraint was ACTIVITY, then the constraint value is
+   * set to activity_coefficient * molality_provided_to_this_method
+   * - Note the possibilities for ignoring the values provided to this method mentioned in the
+   * preceeding paragraphs (setting to zero for equilibrium minerals and gases, and constraints
+   * overriding the values provided)!
+   */
+  void setSolventMassAndFreeMolalityAndMineralMolesAndSurfacePots(
```

Comment:
Talk about a handful

Reply:
Yep, super long.  Evidently i wanted to be really clear what was going on in this method.